models/src/main/scala/coop/rchain/models/rholang/sort/GroundSortMatcher.scala

```diff
@@ -0,0 +1,78 @@
+package coop.rchain.models.rholang.sort
+
+import coop.rchain.models.Expr.ExprInstance
+import coop.rchain.models.Expr.ExprInstance._
+import coop.rchain.models.{KeyValuePair, Par, SortedHashSet}
+import cats.implicits._
+import cats.syntax._
+import coop.rchain.models.rholang.implicits._
+import coop.rchain.models.rholang.sort.ordering._
+
+object GroundSortMatcher {
+  def sortMatch(g: ExprInstance): Either[Throwable, ScoredTerm[ExprInstance]] =
+    g match {
+      case gb: GBool   => ScoredTerm(g, BoolSortMatcher.sortMatch(gb).score).asRight[Throwable]
+      case gi: GInt    => ScoredTerm(g, Leaves(Score.INT, gi.value)).asRight[Throwable]
+      case gs: GString => ScoredTerm(g, Node(Score.STRING, Leaf(gs.value))).asRight[Throwable]
+      case gu: GUri    => ScoredTerm(g, Node(Score.URI, Leaf(gu.value))).asRight[Throwable]
+      case EListBody(gl) =>
+        gl.ps.toList
+          .traverse(par => ParSortMatcher.sortMatch(par))
+          .map(pars =>
+            ScoredTerm(EListBody(gl.withPs(pars.map(_.term.get))),
+                       Node(Score.ELIST, pars.map(_.score): _*)))
+      case ETupleBody(gt) =>
+        gt.ps.toList
+          .traverse(par => ParSortMatcher.sortMatch(par))
+          .map(pars =>
+            ScoredTerm(ETupleBody(gt.withPs(pars.map(_.term.get))),
+                       Node(Score.ETUPLE, pars.map(_.score): _*)))
+      // Note ESet and EMap rely on the stableness of Scala's sort
+      // See https://github.com/scala/scala/blob/2.11.x/src/library/scala/collection/SeqLike.scala#L627
+      case ESetBody(gs) =>
```

Comment:
We could get rid of that `case` clause if `SortedHashSet` was keeping the tree of `ScoredTerm`s but I am not convinced yet.